src/tbb/arena.cpp

```diff
@@ -274,80 +340,94 @@ bool arena::has_enqueued_tasks() {
     return !my_fifo_task_stream.empty();
 }
 
-bool arena::is_out_of_work() {
-#if __TBB_ENQUEUE_ENFORCED_CONCURRENCY
-    if (my_local_concurrency_flag.try_clear_if([this] {
-        return !has_enqueued_tasks();
-    })) {
-        my_market->adjust_demand(*this, /* delta = */ -1, /* mandatory = */ true);
+void arena::request_workers(int mandatory_delta, int workers_delta, bool wakeup_threads) {
+    my_threading_control->adjust_demand(my_tc_client, mandatory_delta, workers_delta);
+
+    if (wakeup_threads) {
+        // Notify all sleeping threads that work has appeared in the arena.
+        get_waiting_threads_monitor().notify([&] (market_context context) {
+            return this == context.my_arena_addr;
+        });
     }
```

Comment:
I would think that this part better belong to threading control rather than arena itself. Moreover, if we looked at the implementation of the [`get_waiting_threads_monitor()`](https://github.com/oneapi-src/oneTBB/pull/758/files#diff-277ea5199fd9ffb60eb8432bd91d0b65b7b58bb13ea06e0856b4f5330d337d56R427), we would see that it is a getter from the `threading_control`.
What do you think if we move it to the `threading_control` entity?

Also, is not that waking up the threads decision can be made based on the difference between current and desired number of the worker threads? I mean we should notify iff current < desired, should not we? Are there other conditions for notification?